quality_of_service_demo/src/liveliness.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rcutils/cmdline_parser.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "quality_of_service_demo/common_nodes.hpp"
+
+static const char * OPTION_POLICY = "--policy";
+static const char * DEFAULT_POLICY = "AUTOMATIC";
+static const char * OPTION_NODE_ASSERT_PERIOD = "--node-assert-period";
+static const char * OPTION_TOPIC_ASSERT_PERIOD = "--topic-assert-period";
+static const char * OPTION_KILL_PUBLISHER_AFTER = "--kill-publisher-after";
+static const size_t DEFAULT_KILL_PUBLISHER_AFTER = 3000;
+
+void print_usage()
+{
+  printf("Usage for liveliness:\n");
+  printf("liveliness lease_duration [-p liveliness_policy] [-n node_assert_liveliness_period] "
+    "[-t topic_assert_liveliness_period] [-h]\n");
+  printf("required arguments:\n");
+  printf("lease_duration: Duration (in ms) before an inactive Publisher is considered not-alive. "
+    "0 means infinity.\n");
+  printf("options:\n");
+  printf("-h : Print this help message.\n");
+  printf("%s liveliness_policy : You may specify AUTOMATIC, MANUAL_BY_NODE, or MANUAL_BY_TOPIC. "
+    "Defaults to %s\n", OPTION_POLICY, DEFAULT_POLICY);
+  printf("%s node_assert_period : How often the Publisher will assert the liveliness of its "
+    "Node (in uint milliseconds). Defaults to 0 (never)\n", OPTION_NODE_ASSERT_PERIOD);
+  printf("%s topic_assert_period : How often the Publisher will assert the liveliness of its "
+    "Publisher (in uint milliseconds). Defaults to 0 (never)\n", OPTION_TOPIC_ASSERT_PERIOD);
+  printf("%s kill_publisher_after : "
+    "Kill the publisher after this amount of time (in uint milliseconds). "
+    "In AUTOMATIC - destroy the whole node. "
+    "In MANUAL_BY_NODE, stop node liveliness assertion. "
+    "In MANUAL_BY_TOPIC, stop topic liveliness assertion. "
+    "Defaults to %lu\n",
+    OPTION_KILL_PUBLISHER_AFTER, DEFAULT_KILL_PUBLISHER_AFTER);
+}
+
+class LivelinessDemo
```

Comment:
Why is this one a class where the other two happen all in the main function?

For consistency, I'd recommend using just one style, probably the main function one since it's easier to see all the variables in scope and get a sense for all the moving parts (rather than having them declared in the private section of the class and using in the constructor). This really doesn't need to be a class, there's no destructor (members manage their own lifecycles), and the methods could just as easily be functions or lambda's.

Reply:
Ok I'll unwrap it